feat(ray): Implement dynamic scale-in for RaySwordfishActor

huleilei opened the pull request:
## Changes Made
This commit implements the dynamic scaling down (scale-in) functionality for RaySwordfishActor to release idle resources.

- **Idle Worker Retirement**: Implemented `retire_idle_ray_workers` in `RayWorkerManager`. It identifies workers that have been idle for longer than `DAFT_AUTOSCALING_DOWNSCALE_IDLE_SECONDS` (default: 60s) and releases them.
- **Scheduler Integration**: The `SchedulerActor` loop now periodically checks for idle workers and triggers retirement while maintaining a minimum survivor count (`DAFT_AUTOSCALING_MIN_SURVIVOR_WORKERS`, default: 1).
- **Blacklist Mechanism**: Added a `pending_release_blacklist` to prevent the autoscaler from immediately respawning workers that were just released.

## Configuration
New environment variables added:
- `DAFT_AUTOSCALING_DOWNSCALE_ENABLED`: Enable/disable downscaling (default: `true`).
- `DAFT_AUTOSCALING_DOWNSCALE_IDLE_SECONDS`: Seconds a worker must be idle before retirement (default: `60`).
- `DAFT_AUTOSCALING_MIN_SURVIVOR_WORKERS`: Minimum number of workers to keep alive (default: `1`).
- `DAFT_AUTOSCALING_PENDING_RELEASE_EXCLUDE_SECONDS`: TTL for blacklisted worker IDs (default: `120`).
- 


<!-- Describe what changes were made and why. Include implementation details if necessary. -->

## Related Issues
https://github.com/Eventual-Inc/Daft/issues/5683
<!-- Link to related GitHub issues, e.g., "Closes #123" -->

srilman reviewed (commented):
Overall this makes sense to me and I think would be useful. However, I feel like it would make more sense for the worker manager to have the logic to determine what workers should be retired vs within the scheduler and then passing into the worker manager.

madvart commented:
@huleilei - Checking to see if you are ok working on changes recommended by @srilman

huleilei commented:
I'm

> @huleilei - Checking to see if you are ok working on changes recommended by @srilman

Sorry, I have been reviewing this PR recently. Thank you

huleilei commented:
@srilman @madvart Please help me review when you have time, thanks.

DogerW666 commented:
This PR is also useful for us.

srilman commented on src/daft-distributed/src/scheduling/scheduler/scheduler_actor.rs:
```diff
@@ -138,6 +211,31 @@ where
+                            "Downscale: retired idle workers"
+                        );
+                    }
+                }
```
Rather than having this code here, can we move it to the WorkerManager directly? Plus, rather than a specific `retire_idle_ray_workers`, can we have a generic `retire_idle_workers`

srilman reviewed (approved):
Overall, LGTM. Had 1 small thing, but once thats addressed plus the merge conflict, we can merge

madvart commented:
@huleilei - Thanks for your patience and work on this. Once you are able to resolve the last comment and the merge conflicts, we can merge this. Thanks again.

huleilei committed:
feat(ray): add opt-in downscaling for Ray autoscaler

Adds an opt-in Ray downscaling (scale-in) mechanism by retiring idle Flotilla workers to help Ray autoscaler shrink clusters when workloads become idle.

Highlights:
- Retire idle Ray workers with configurable idle threshold and a min-survivor floor.
- Head-node protection and a pending-release blacklist to avoid immediate respawn.
- Expose configuration via `daft.set_runner_ray(...)` and environment variables.
- Keep scale-up behavior aligned with upstream high-water-mark ramp-up logic.
- Document autoscaling/downscaling in `docs/distributed/ray.md`.

Follow-ups on top of upstream PR https://github.com/Eventual-Inc/Daft/pull/5903:
- Resolve merge conflict with `main`: adopt the new `(scheduled, cancelled)` tuple from `Scheduler::schedule_tasks` and the `worker_id` field on `TaskEvent::Scheduled`.
- Address @srilman review: move the per-tick downscale gating from `scheduler_actor.rs` into the generic `WorkerManager::retire_idle_workers`; `RayWorkerManager` now owns the enable flag, min-survivor floor, idle thresholds, head-node protection and blacklist TTLs.

huleilei commented:
@madvart @srilman Hello, I have made the revisions. Thank you for reviewing them again. Best regards

huleilei committed:
Merge branch 'main' into hll/auto

huleilei commented:
@madvart @srilman Hello, I have made the revisions. Thank you for reviewing them again. Best regards
